### ISSUE TITLE
Unexpected behavior on `dart(prefer_const_declarations)` rule on Windows 10

### ISSUE THREAD

**Issue by lakscastro** [dart-lang/linter#3053]
**Describe the issue**

Analyzer don't trigger `dart(prefer_const_declarations)` on Windows if using `dart:math` on a declaration.

**Printscreens**

![image](https://user-images.githubusercontent.com/51419598/140029993-9d3caf84-afa9-40b8-a9ec-0c9f0fbf0d15.png)

![image](https://user-images.githubusercontent.com/51419598/140030060-f2f3c536-65f8-43c8-a6c5-f54dedeaa965.png)

**Examples**

```dart
import 'dart:math' as math;

void main() {
  /// Trigger `Prefer const over final for declarations.`  Working ✔️
  final a = 1 + 1;

  /// Don't trigger nothing even `math.pi` being a constant variable ❌
  /// Expected behavior: Trigger the rule `Prefer const over final for declarations.`
  final b = 1 + 1 + math.pi;
}
```

**Rule: `dart(prefer_const_declarations)`**

**To Reproduce**
Ideally, please include an example.

## [Reproducible Dart Project Zipped Folder](https://github.com/dart-lang/linter/files/7466248/dartpad.zip)

You can download the zipped project above or just create a project with the configuration described in this issue.

### Steps:

1. Install packages `dart pub get`
2. Open `lib/main.dart`

**Expected behavior**

The following:

```dart
final a = 1 + 1 + math.pi; 
```

Should trigger the `Prefer const over final for declarations.` rule because this is a constant declaration in a final declaration.

**Additional context**

### Flutter Context
```
Flutter 2.5.3 • channel stable • https://github.com/flutter/flutter.git
Framework • revision 18116933e7 (3 weeks ago) • 2021-10-15 10:46:35 -0700
Engine • revision d3ea636dc5
Tools • Dart 2.14.4
```

### Linter Version and Reproducible Project Context

```yaml
name: dartpad
description: DartPad
publish_to: "none"
version: 1.0.0+1

environment:
  sdk: ">=2.14.0 <3.0.0"
  flutter: "2.5.3"

dev_dependencies:
  flutter_lints: ^1.0.4
```

### Analysis Options

```yaml
include: package:flutter_lints/flutter.yaml

linter:
  rules:
```

**Comment by pq** [comment]
Thanks for the thoughtful report and repro.  Interestingly this does not reproduce locally on Mac OS with a recent SDK build.  I suppose it *could* be Windows-related though I'd be surprised.  More likely I wonder if there have been improvements in constant evaluation that could account for why it's not reproducing on a fresh SDK.

/cc @scheglov @srawlins

**Comment by srawlins** [comment]
I made changes recently to const code, but I would be surprised if it affected a simple case like this.

**Comment by LaksCastro** [comment]
I know that this happens only windows beause I use a CI validator, and the CI is using `macos-latest` so the CI triggered errors that don't happened on my local machine (windows) after searching a while I noticed that the windows is not triggering any constant (with `final` or `var`) with another const variable of _another package._ But I'll investigate further and try to provide more details.

**Comment by pq** [comment]
Ah.  Fantastic.  Thanks so much!

**Comment by scheglov** [comment]
Might be the same issue as to be fixed with https://dart-review.googlesource.com/c/sdk/+/218983
We did not evaluate dependencies of the expression being evaluated.
So, if `pi` has not been used yet, its value is unknown, and the whole expression is invalid.

**Comment by LaksCastro** [comment]
I created a reproducible repository here [dart-linter-issue-reproducible-step.](https://github.com/LaksCastro/dart-linter-issue-reproducible-step)

This repo uses Github Actions to perform `dart analyze`

You can see at pull request tabs that doesn't trigger anything even at `main.dart` is missing a const variable. 

- [MacOS machine - macOS Big Sur 11](https://github.com/LaksCastro/dart-linter-issue-reproducible-step/pull/2/files)
- [Windows machine - Windows Server 2019](https://github.com/LaksCastro/dart-linter-issue-reproducible-step/pull/1)

**Comment by LaksCastro** [comment]
I hadn't seen @‎scheglov comment, but I think that's the reason.

**Comment by pq** [comment]
Awesome.  Thanks for the follow-up!

**Comment by scheglov** [comment]
Landed as https://github.com/dart-lang/sdk/commit/c6fe2b6f6a39ca3865559f6c88dbadeb62410896.

**Comment by lakscastro** [comment]
I guess we can set this issue as solved! Great work 

<kbd>
  <img src="https://user-images.githubusercontent.com/51419598/156901980-2cbf0e9e-a3bf-408f-a3bc-154c4377004a.png">
</kbd>

**Comment by pq** [comment]
Thanks for confirming @lakscastro!
